httpd/handler.go

```diff
@@ -129,8 +132,33 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		}
 		rp.ServeHTTP(w, r)
 		return
+	} else if mount.LocalDir != "" {
+		path := filepath.Join(mount.LocalDir, r.URL.Path)
+
+		if mount.AppendSuffix {
+			path = filepath.Join(mount.LocalDir, strings.TrimPrefix(r.URL.Path, mount.Path))
+		}
```

Comment:
After joining the paths (which implicitly cleans the result), you need to check, again that the Path is within the base root allowed. Otherwise, using a path such that `../../../../etc/shadow` would escape the base root.

"Cleaning" gets rid of the relative `..`, so a simple prefix match should be enough after this.

Reply:
Ah I see, did not think of this.

Reply:
Wait, I'm a little confused. What value will be `../../../etc/shadow`? If its URL path, it will be `mount.LocalDir/etc/shadow` at the moment right? I can't picture the scenario where it will be outside of localDir.

Reply:
Imagine `path: /subdir`, `localDir: /tftproot` and query URL: `/subdir/../etc/shadow`, then after trimming `/subdir`, you get `'../etc/shadow`, then you concat that with root: `/tftproot/../etc/shadow`, when you open that file, you're actually opening `/etc/shadow`. Am I right?

Reply:
https://play.golang.org/p/nrP2nyT2K0W 🤦🏾 

I was under the impression that this would actually still be `tftpboot/test` but it's not. So you are indeed right! So your suggestion is the following: 

```suggestion
		path := filepath.Join(mount.LocalDir, r.URL.Path)

		if mount.AppendSuffix {
			path = filepath.Join(mount.LocalDir, strings.TrimPrefix(r.URL.Path, mount.Path))
			if !strings.HasPrefix(path, mount.LocalDir) {
				h.server.logger.Error().
					Err(err).
					Msgf("Requested path is invalid", path)
				http.Error(w, err.Error(), http.StatusBadRequest)
				return
			})
		}
```

Reply:
I'd do this:

```suggestion
		path := filepath.Join(mount.LocalDir, mount.Path)
		
		if mount.AppendSuffix {
			path = filepath.Join(mount.LocalDir, strings.TrimPrefix(r.URL.Path, mount.Path))
		}
		
		if !strings.HasPrefix(path, mount.LocalDir) {
			h.server.logger.Error().
				Err(err).
				Msgf("Requested path is invalid", path)
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		})
```

Note that without `AppendSufix`, the path is NOT touching the `r.URL.Path` (no need to), instead it just takes the path from manifest.